src/watchman_client.js

```
@@ -0,0 +1,341 @@
+    var self = this;
+
+    let watcherInfo = createWatcherInfo(watchmanWatcher);
+    let subscription = watcherInfo.subscription;
+    
+    let promise = new Promise((resolve, reject) => {
+      watch(subscription, root)
+        .then((resp) => clock(subscription))
+        .then((data) => {
+            let watcherInfo = getWatcherInfo(subscription);
+
+            // create the 'bare' options w/o 'since' or relative_root.
+            // Store in watcherInfo for later use if we need to reset
+            // things after an 'end' caught here.
+            let options = watchmanWatcher.createOptions();  
+            watcherInfo.options = options;
+            
+            // Dup the options object so we can add 'relative_root' and 'since'
+            // and leave the original options object alone. We'll do this again
+            // later if we need to resubscribe after 'end' and reconnect.
+            options = Object.assign({}, options);
+
+            if (_relative_root) {
+              options.relative_root = watcherInfo.relativePath;
+            }
+
+            options.since = watcherInfo.since;
+
+            _client.command(['subscribe',
+                             watcherInfo.root,
+                             subscription,
+                             options],
+                            (error, resp) => {
+                              if (error) {
+                                reject(error);
+                              } else {
+                                resolve(resp);
+                              }
+                            });
+            console.log("For subscription '" + subscription +
+                        "', subscribing against root '" + watcherInfo.root +
+                        "', with relative_root = '" + options.relative_root + "'");
+        })
+        .catch(function(error) {
+          console.error("Caught error in subscribe for watchmanWatcher root " + root);
```

Unless there is a really strong motivation, I would prefer if libraries don't log things  by default. Otherwise log output gets unwieldy.

If there is strong motivation let me know, otherwise I suspect we may not want to alwayas log this, rather using something like https://www.npmjs.com/package/debug

---

Stef, the error indicates that a subscription failed. Isn't that something that should be reported? (In general I agree with moving most things to debug, but not sure on this one).

---

Comments on my comment? I could see putting something in WatchmanWatcher and removing it from WatchmanClient, but I do think that swallowing the error (by just rejecting and not putting anything out at all) isn't a great solution. Comments?